Source/WebCore/html/HTMLMetaElement.cpp

```diff
@@ -106,11 +106,19 @@ void HTMLMetaElement::attributeChanged(const QualifiedName& name, const AtomStri
         if (isInDocumentTree()) {
             if (equalLettersIgnoringASCIICase(oldValue, "theme-color"_s) && !equalLettersIgnoringASCIICase(newValue, "theme-color"_s))
                 document().metaElementThemeColorChanged(*this);
+#if ENABLE(DARK_MODE_CSS)
+            else if (equalLettersIgnoringASCIICase(oldValue, "color-scheme"_s) || equalLettersIgnoringASCIICase(oldValue, "supported-color-schemes"_s))
+                document().metaElementColorSchemeChanged();
+#endif
         }
         break;
     case AttributeNames::contentAttr:
         m_contentColor = std::nullopt;
         process();
+#if ENABLE(DARK_MODE_CSS)
+        if (const AtomString& nameValue = attributeWithoutSynchronization(nameAttr); equalLettersIgnoringASCIICase(nameValue, "color-scheme"_s) || equalLettersIgnoringASCIICase(nameValue, "supported-color-schemes"_s))
+            document().metaElementColorSchemeChanged();
+#endif
```

Comment:
This code is sort of duplicated with what's inside the `process()` function. Except that method calls the outdated `document().processColorScheme()` method (which should probably be cleaned up/removed)

Reply:
Following [268064@main](https://commits.webkit.org/268064@main), `process()` calls `metaElementColorSchemeChanged()`, so I'm wondering why this part of the change is necessary?

Reply:
> This code is sort of duplicated with what's inside the `process()` function. Except that method calls the outdated `document().processColorScheme()` method

`process()` on the main branch (and in this PR branch too) no longer calls `document().processColorScheme()` directly, right? Instead `process()` also now calls `metaElementColorSchemeChanged()`

>  (which should probably be cleaned up/removed)

When I looked at `document().processColorScheme()` and `processColorSchemeString()`(and talked to you about them), it had seemed to me that it was doing some parsing that was redundant with `CSSParser::parseSingleValue(CSSPropertyColorScheme,…)` — but after looking back at them further, I can see now that they’re not really re-parsing the value but instead are splitting the parsed color scheme (in the case where the color scheme contains multiple values), and then doing necessary processing based on the value.

But maybe I’m misunderstanding something?

Reply:
> Following [268064@main](https://commits.webkit.org/268064@main), `process()` calls `metaElementColorSchemeChanged()`, so I'm wondering why this part of the change is necessary?

(Apologies in advance if I’m misunderstanding which exact part you mean, but to be clear, I’ll attempt to write out here a prose explanation of what the code is attempting to do, and why —

`process()` doesn’t handle the case where the `name` or `content` attributes are _removed_; `process()` only looks at the _existing_ values of those attributes that it finds in the current DOM.

So, to describe explicitly the two things this patch is doing:
- In the case where a `name` attribute has changed — including the case where the `name` attribute has been completely removed (which `process()` doesn’t handle) — this patch checks the _old_ value of the `name` attribute, and if that _was_ `color-scheme`, then calls `document().metaElementColorSchemeChanged()`.
- In the case where a `content` attribute has changed — including the case where the `content` attribute has been completely removed (which `process()` doesn’t handle) — this patch checks the current value of the `name` attribute, and if that is `color-scheme`, then  calls `document().metaElementColorSchemeChanged()`.

If it seems like it might help to add comments to the code, please let me know, and I’ll do that.

In terms of test cases, without this patch, the following two asserts from http://wpt.live/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-attribute-changes.html fail:
- _Removed name attribute from meta color-scheme_
- _Removed the content attribute of meta color-scheme_

…Whereas, with this patch, those two asserts now pass.

Reply:
@pxlcoder, @nt1m Please lemme know if there’s anything more I need to do here at this point

Reply:
> `process()` doesn’t handle the case where the name or content attributes are removed; `process()` only looks at the existing values of those attributes that it finds in the current DOM.

I think we might want to restructure that behavior? Currently, this code would result in scenarios where `metaElementColorSchemeChanged()` is called twice.

The comment in `process()` says: "// All below situations require a content attribute (which can be the empty string)." But that doesn't actually appear to be true for all of them.

Reply:
> Currently, this code would result in scenarios where `metaElementColorSchemeChanged()` is called twice.

Ah yeah — OK, I’ve pushed an update which attempts to address that. Please let me know if you don’t think it’s the right way to deal with it